qualtran/qref_interop/_bloq_to_qref.py

```diff
@@ -0,0 +1,311 @@
+import networkx as nx
+
+from functools import singledispatch
+from typing import Any, Iterable, Optional, Union
+
+import sympy
+from qref.schema_v1 import PortV1, RoutineV1, SchemaV1
+
+from qualtran import Bloq, BloqInstance, CompositeBloq
+from qualtran import Connection as QualtranConnection
+from qualtran import Register, Side, Soquet
+from qualtran.cirq_interop import CirqGateAsBloq
+
+
+@singledispatch
+def _bloq_type(bloq: Bloq) -> str:
+    """Determine type of Bloq.
+
+    Output of this function is used for populating the `type` field of RoutineV1
+    build from imported Bloqs. By default, we use name of the type of the Bloq.
+    """
+    return type(bloq).__name__
+
+
+@_bloq_type.register
+def _cirq_gate_bloq_type(bloq: CirqGateAsBloq) -> str:
+    """Determine type of Bloq constructed from Cirq gate.
+
+    Without this variant of _bloq_type, type of all instances of CirqGateAsBloq
+    are set to "CirqGateAsBloq" which is not helpful, because all gates would get the
+    same type. Instead of this default behaviour, we extract textual representation
+    of the Cirq's gate.
+    """
+    return str(bloq.gate)
+
+
+def _is_symbol_or_int(expression):
+    try:
+        int(expression)
+        return True
+    except (TypeError, ValueError):
+        return expression.isidentifier()
+
+
+def _extract_common_bloq_attributes(bloq: Bloq, name: Optional[str] = None) -> dict[str, Any]:
+    """Extract common bloq attributes such as name, type and ports.
+
+    There are several Bloq classes, however, they all share common set of atributes.
+    This function is used to extract them, so that we don't have to duplicate logic
+    for each Bloq subtype.
+
+    Args:
+        bloq: Bloq for which information should be extracted.
+        name: Optional override for name. Setting it to None will cause some reasonable
+            name (depending on bloq type) to be inferred. The typical use of non-trivial
+            name is when a better name is known from the parent Bloq.
+
+    Returns:
+        A dictionary that can be unpacked into arguments of RoutineV1 initializer.
+    """
+    ports = [port for reg in bloq.signature for port in _ports_from_register(reg)]
+    local_variables = {}
+    for port in ports:
+        if not _is_symbol_or_int(str(port.size)):
+            local_variable_name = f"{port.name}_size"
+            local_variables[local_variable_name] = port.size
+            port.size = local_variable_name
+
+    if name is None:
+        name = bloq.__class__.__name__
+        try:
+            if bloq.uncompute:
+                name += "_uncompute"
+        except AttributeError:
+            pass
+        try:
+            if bloq.is_adjoint:
+                name += "_adjoint"
+        except AttributeError:
+            pass
+    input_params = _extract_input_params(bloq, ports, local_variables)
```

Comment:
Should `_extract_input_params` just deal with `bloq` and `ports` and the `local_variables` should be subtracted in this function because it's specific to the mapping of new port sizes to old port sizes done here?